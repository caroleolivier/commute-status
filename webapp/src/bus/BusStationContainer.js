import 'whatwg-fetch';
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import BusStationHeader from './BusStationHeader';
import BusArrivalTimeTable from './BusArrivalTimeTable';
import Loading from '../common/Loading';
import ErrorMessage from '../common/ErrorMessage';
import TfLBusDataAPIService from './services/TfLBusDataAPIService';

const BusStationContainerState = {
    LOADING: {
        id: 0,
        getComponent: () => <Loading />
    },
    ERROR: {
        id: 1,
        getComponent: () => <ErrorMessage message="Unable to load bus data" />
    },
    LOADED: {
        id: 2,
        getComponent: container => <BusArrivalTimeTable arrivals={container.state.busesExpectedTimes} />
    }
};

class BusStationContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            state: BusStationContainerState.LOADING,
            busesExpectedTimes: []
        };
    }

    componentDidMount() {
        const service = new TfLBusDataAPIService();
        service.fetch(this.props.busStation.stationId,
            this.props.busStation.directionId, this.props.busStation.buses)
            .then((busArrivalTimes) => {
                /* lot of discussions around whether setState should be called here or not :s */
                /* eslint-disable react/no-did-mount-set-state */
                this.setState({
                    busesExpectedTimes: busArrivalTimes,
                    state: BusStationContainerState.LOADED
                });
                /* eslint-enable react/no-did-mount-set-state */
            })
            .catch((ex) => {
                console.log(ex);
                this.setState({
                    state: BusStationContainerState.ERROR
                });
            });
    }

    render() {
        return (
            <div>
                <BusStationHeader
                    stationName={this.props.busStation.stationName}
                    direction={this.props.busStation.direction}
                />
                {this.state.state.getComponent(this)}
            </div>
        );
    }
}

BusStationContainer.propTypes = {
    busStation: PropTypes.shape({
        stationName: PropTypes.string,
        stationId: PropTypes.string,
        direction: PropTypes.string,
        directionId: PropTypes.string,
        buses: PropTypes.arrayOf(PropTypes.string)
    }).isRequired
};

export default BusStationContainer;

export { BusStationContainerState };
