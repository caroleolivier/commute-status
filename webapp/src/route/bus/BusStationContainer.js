import 'whatwg-fetch';
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import BusStationHeader from './BusStationHeader';
import BusArrivalTimeTable from './BusArrivalTimeTable';
import Loading from '../common/Loading';
import ErrorMessage from '../common/ErrorMessage';
import TfLDataAPIService from '../../services/TfLDataAPIService';

import styles from './styles.scss';

const BusStationContainerState = {
    LOADING: {
        id: 0,
        getComponent: () => <Loading />
    },
    ERROR: {
        id: 1,
        getComponent: () => <ErrorMessage message="Unable to load bus data" />
    },
    LOADED: {
        id: 2,
        getComponent: container => <BusArrivalTimeTable arrivals={container.state.busesExpectedTimes} />
    }
};

class BusStationContainer extends Component {
    constructor(props) {
        super(props);
        this.state = {
            state: BusStationContainerState.LOADING,
            busesExpectedTimes: []
        };
    }

    componentDidMount() {
        const service = new TfLDataAPIService();
        service.fetchArrivals(this.props.config.stationId,
            this.props.config.directionId, this.props.config.lines)
            .then((busArrivalTimes) => {
                /* lot of discussions around whether setState should be called here or not :s */
                /* eslint-disable react/no-did-mount-set-state */
                this.setState({
                    busesExpectedTimes: busArrivalTimes,
                    state: BusStationContainerState.LOADED
                });
                /* eslint-enable react/no-did-mount-set-state */
            })
            .catch((ex) => {
                console.log(ex);
                this.setState({
                    state: BusStationContainerState.ERROR
                });
            });
    }

    render() {
        return (
            <div className={styles.container}>
                <BusStationHeader
                    stationName={this.props.config.stationName}
                    direction={this.props.config.direction}
                />
                {this.state.state.getComponent(this)}
            </div>
        );
    }
}

BusStationContainer.propTypes = {
    config: PropTypes.shape({
        stationName: PropTypes.string,
        stationId: PropTypes.string,
        direction: PropTypes.string,
        directionId: PropTypes.string,
        lines: PropTypes.arrayOf(PropTypes.string)
    }).isRequired
};

export default BusStationContainer;

export { BusStationContainerState };
