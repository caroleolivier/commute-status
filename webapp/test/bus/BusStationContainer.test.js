/* eslint-env browser, jest */
import fetchMock from 'fetch-mock';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import BusStationContainer from '../../src/bus/BusStationContainer';

describe('BusStationContainer component', () => {
    const busStation = {
        stationName: 'Bla Street',
        stationId: '49030283ID',
        direction: 'Waterloo',
        directionId: 'outbound',
        buses: ['10', 'N20']
    };
    let busStationContainer;
    let fetchDataObj;
    let promise;

    describe('Given component was mounted successfully', () => {
        beforeEach(() => {
            promise = new Promise((resolve, reject) => {
                fetchDataObj = { resolve, reject };
            });
            fetchMock.get('*', promise);
            busStationContainer = shallow(<BusStationContainer busStation={busStation} />);
        });

        afterEach(() => {
            fetchMock.restore();
        });

        test('it displays a loading state', () => {
            const tree = toJson(busStationContainer);
            expect(tree).toMatchSnapshot('loading');
        });

        test('it should send a request to load bus expected arrivals', () => {
            fetchMock.called('https://api.tfl.gov.uk/Line/10%2CN20/Arrivals/49030283ID?direction=outbound');
        });

        describe('Given the server request was successful', () => {
            const busArrivals = [
                { lineName: '10', stationName: 'Bla Street', timeToStation: 60 },
                { lineName: '10', stationName: 'Bla Street', timeToStation: 90 },
                { lineName: '20', stationName: 'Bla Street', timeToStation: 30 }
            ];

            beforeEach(() => {
            });

            test('it should load the bus expected arrivals', () => {
                fetchDataObj.resolve(busArrivals);
                busStationContainer.update();
                console.log('testing')
                const tree = toJson(busStationContainer);
                expect(busStationContainer.instance().state.state).toEqual('loaded');
                expect(tree).toMatchSnapshot('loaded');
            });
        });

        describe('Given the server request failed', () => {
            test('it should display an error message', () => {

            });
        });
    });
});


// describe('BusStationContainer component', () => {
//     const busArrivals = [
//         { lineName: '10', stationName: 'Bla Street', timeToStation: 60 },
//         { lineName: '10', stationName: 'Bla Street', timeToStation: 90 },
//         { lineName: '20', stationName: 'Bla Street', timeToStation: 30 }
//     ];
//     beforeEach(() => {
//         fetchMock.get('https://api.tfl.gov.uk/Line/10%2CN20/Arrivals/49030283ID?direction=outbound',
//             busArrivals);
//     });

//     afterEach(() => {
//         fetchMock.restore();
//     });

//     test('shallow render matches snapshot.', () => {
//         const busStation = {
//             stationName: 'Bla Street',
//             stationId: '49030283ID',
//             direction: 'Waterloo',
//             directionId: 'outbound',
//             buses: ['10', 'N20']
//         };
//         const busStationContainer = shallow(<BusStationContainer busStation={busStation} />);
//         const tree = toJson(busStationContainer);
//         expect(tree).toMatchSnapshot('rendering');
//     });
// });
