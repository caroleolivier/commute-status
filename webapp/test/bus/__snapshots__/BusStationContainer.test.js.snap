// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading 1`] = `
<div>
  <BusStationHeader
    direction="Waterloo"
    stationName="Bla Street"
  />
  <p>
    Loading...
  </p>
</div>
`;
